### PR TITLE
Seq.indexOf(element,start); Seq.lastIndexOf(element,end)

### PR DESCRIPTION
**Pull request by ruslansennov**
+ unit tests
+ missing unit tests

### CODE REVIEW

File: src/main/java/javaslang/collection/Stream.java [hunk 1]
```diff
@@ -927,10 +927,10 @@ default int indexOf(T element) {
     Stream<T> intersperse(T element);
 
     @Override
-    default int lastIndexOf(T element) {
+    default int lastIndexOf(T element, int end) {
         int result = -1, index = 0;
         for (Stream<T> stream = this; !stream.isEmpty(); stream = stream.tail(), index++) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Stream.java]: Maybe the for condition `!stream.isEmpty()` can be extended to `index <= end && !stream.isEmpty()`? ...

File: src/main/java/javaslang/collection/Stream.java [hunk 2]
```diff
@@ -927,10 +927,10 @@ default int indexOf(T element) {
...
-            if (Objects.equals(stream.head(), element)) {
+            if (index <= end && Objects.equals(stream.head(), element)) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Stream.java]: ... then we can remove the check here and the loop ends, if `index > end`.

Would this be the same semantic?

File: src/main/java/javaslang/collection/Queue.java [hunk 3]
```diff
@@ -645,13 +645,13 @@ public T head() {
...
-    public int indexOf(T element) {
-        final int frontIndex = front.indexOf(element);
+    public int indexOf(T element, int from) {
+        final int frontIndex = front.indexOf(element, from);
         if (frontIndex != -1) {
             return frontIndex;
         } else {
             // we need to reverse because we search the first occurrence
-            final int rearIndex = rear.reverse().indexOf(element);
+            final int rearIndex = rear.reverse().indexOf(element, from - front.length());
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Queue.java]: Did we check the case that `front.length() > from`. Just asking for the case ... :) I didn't dig deep into it...

File: src/main/java/javaslang/collection/List.java [hunk 4]
```diff
@@ -839,10 +839,10 @@ default int indexOf(T element) {
     }
 
     @Override
-    default int lastIndexOf(T element) {
+    default int lastIndexOf(T element, int end) {
         int result = -1, index = 0;
         for (List<T> list = this; !list.isEmpty(); list = list.tail(), index++) {
-            if (Objects.equals(list.head(), element)) {
+            if (index <= end && Objects.equals(list.head(), element)) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/List.java]: See my comment at Stream.java below (for-loop)

File: src/main/java/javaslang/collection/Stream.java (hunk 1 above)
Comment (ruslansennov) [inline src/main/java/javaslang/collection/Stream.java]: Of course you're right, shame on me :)

File: src/main/java/javaslang/collection/Queue.java (hunk 3 above)
Comment (ruslansennov) [inline src/main/java/javaslang/collection/Queue.java]: I believe it's all right here... we just continue searching in rear part of queue with start < 0

Comment (danieldietrich) [inline src/main/java/javaslang/collection/Queue.java]: ok, thx!

File: src/main/java/javaslang/collection/Stream.java (hunk 1 above)
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Stream.java]: Don't shame on you - we team up to get the best result. There is nearly always room for improvement.

File: src/main/java/javaslang/collection/List.java (hunk 4 above)
Comment (ruslansennov) [inline src/main/java/javaslang/collection/List.java]: fixed in same commit

**Comment by danieldietrich** [comment]
Great work!

**Pull request merged**